algebird-core/src/main/scala/com/twitter/algebird/Monoid.scala

```diff
@@ -47,8 +48,14 @@ trait Monoid[@specialized(Int, Long, Float, Double) T] extends Semigroup[T] {
       None
     }
   }
-  // Override this if there is a more efficient means to implement this
-  def sum(vs: TraversableOnce[T]): T = sumOption(vs).getOrElse(zero)
+  override def sum(vs: TraversableOnce[T]): T = sumOption(vs).getOrElse(zero)
+
+  /**
+   * These are from algebra.Monoid
+   */
+  override def additive: AMonoid[T] = this
+  override def empty: T = zero
+  override def combineAll(t: TraversableOnce[T]): T = sum(t)
```

Comment:
Multiplication and "plus" and 1 and "zero" is a valid Monoid. Algebird has
just settled this by saying we always prefer additive so that Monoid can be
a subclass of group.

Not everyone agrees with that design. So in algebra, there is a Monoid that
has a combine method, and in the ring package, there are Additive and
Multiplicative monoids. The difference is only used in spire where the
syntax extensions are enabled. + needs an additive Semigroup. * needs a
multiplicative Semigroup.

To keep comparability with our source code, and interop with
monoids/semigroups from other packages, I think we need to do what I did
(extend Semigroup and AdditiveSemigroup).

On Thursday, June 9, 2016, Alex Levenson <notifications@github.com> wrote:

> In algebird-core/src/main/scala/com/twitter/algebird/Monoid.scala
> <https://github.com/twitter/algebird/pull/523#discussion_r66546774>:
>
> > @@ -47,8 +48,14 @@ trait Monoid[@specialized(Int, Long, Float, Double) T] extends Semigroup[T] {
> >        None
> >      }
> >    }
> > -  // Override this if there is a more efficient means to implement this
> > -  def sum(vs: TraversableOnce[T]): T = sumOption(vs).getOrElse(zero)
> > +  override def sum(vs: TraversableOnce[T]): T = sumOption(vs).getOrElse(zero)
> > +
> > +  /**
> > +   * These are from algebra.Monoid
> > +   */
> > +  override def additive: AMonoid[T] = this
> > +  override def empty: T = zero
> > +  override def combineAll(t: TraversableOnce[T]): T = sum(t)
>
> What's the difference between non-additive Monoid and additive Monoid?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/twitter/algebird/pull/523/files/053fd81dafbd88982cad3b9b5c8839ab16ceb399#r66546774>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAEJdkKu96bh2NZ25MRTFe_VCf1H-veFks5qKLRrgaJpZM4Ieuhu>
> .
>


-- 
P. Oscar Boykin, Ph.D. | http://twitter.com/posco | http://pobox.com/~boykin